library/src/blas1/rocblas_swap.hpp

```diff
@@ -13,35 +13,66 @@ __forceinline__ __device__ __host__ void rocblas_swap_vals(T* x, T* y)
     *x    = tmp;
 }
 
-template <typename T>
-__global__ void rocblas_swap_kernel(rocblas_int n, T* x, rocblas_int incx, T* y, rocblas_int incy)
+template <typename T, typename U>
+__global__ void rocblas_swap_kernel(rocblas_int    n,
+                                    U              xa,
+                                    ptrdiff_t      offsetx,
+                                    rocblas_int    incx,
+                                    rocblas_stride stridex,
+                                    U              ya,
+                                    ptrdiff_t      offsety,
+                                    rocblas_int    incy,
+                                    rocblas_stride stridey)
 {
+    T*        x   = load_ptr_batch(xa, hipBlockIdx_y, offsetx, stridex);
+    T*        y   = load_ptr_batch(ya, hipBlockIdx_y, offsety, stridey);
     ptrdiff_t tid = hipBlockIdx_x * hipBlockDim_x + hipThreadIdx_x;
+
     if(tid < n)
     {
         rocblas_swap_vals(x + tid * incx, y + tid * incy);
     }
 }
```

Comment:
rocblas_swap_kernel:

You can probably skip the template parameter T by using "auto x = load_ptr_batch ..." rather than "T * x = ".
Here, T is doing nothing.

Reply:
Changed it but now I think I preferred the original.   Was simpler to see it is computing a pointer, auto here is more like a trick than helping avoid future changes.   Overuse of auto may make maintenance worse rather than better.

Reply:
Even if written like this ?
    ptrdiff_t tid = hipBlockIdx_x * hipBlockDim_x + hipThreadIdx_x;
    if(tid < n)
    {
        rocblas_swap_vals(load_ptr_batch(xa, hipBlockIdx_y, offsetx + tid * incx, stridex), 
                                      load_ptr_batch(ya, hipBlockIdx_y, offsety +  tid * incy, stridey) );
    }

Reply:
Close but that makes the function call look too crowded  :-)     I'll leave it with auto unless someone else spots an issue.

Reply:
> 
> 
> Changed it but now I think I preferred the original. Was simpler to see it is computing a pointer, auto here is more like a trick than helping avoid future changes. Overuse of auto may make maintenance worse rather than better.

You can document and force it to be a pointer:

```
auto* X = load_ptr_batch...
```

Reply:
`auto` is merely a placeholder; template argument deduction rules do the work. So if:
```
template <typename T>
void foo( T**** X);
````
will work, then
```
auto**** X = ...
```
will work similarly.

Reply:
> 
> 
> rocblas_swap_kernel:
> 
> You can probably skip the template parameter T by using "auto x = load_ptr_batch ..." rather than "T * x = ".

auto* x = load_ptr_batch

Reply:
Good point Lee, that is better.